tests/TestMake.hs

```diff
@@ -36,172 +36,510 @@ timestampB = utcMidnightOnDate 2019 1 2
 timestampC = utcMidnightOnDate 2019 1 3
 timestampD = utcMidnightOnDate 2019 1 4
 
+oneSecond :: Int
+oneSecond = 10 ^ (6::Int) -- microseconds.
+
 spec :: Spec
 spec = do
   let sourcesDir = "tests/purs/make"
   let moduleNames = Set.fromList . map P.moduleNameFromString
+  let modulePath name = sourcesDir </> (T.unpack name <> ".purs")
+  let foreignJsPath name = sourcesDir </> (T.unpack name <> ".js")
+
+  -- Test helpers.
+  let testN fn name modules compile2 res =
+        fn name $ do
+          let names = map (\(mn, _, _) -> mn) modules
+          let paths = map modulePath names
+          let timestamp = utcMidnightOnDate 2019 1
+
+          forM_ (zip [0..] modules) $ \(idx, (mn, content, _)) -> do
+            writeFile (modulePath mn) (timestamp idx) content
+
+          compile paths `shouldReturn` moduleNames names
+
+          forM_ (zip [length modules..] modules) $ \(idx, (mn, _, mbContent)) -> do
+            maybe (pure ()) (writeFile (modulePath mn) (timestamp idx)) mbContent
+
+          compile2 paths `shouldReturn` moduleNames res
+
+  let test2 fn name (mAContent1, mAContent2, mBContent) res =
+        testN fn name
+          [ ("A", mAContent1, Just mAContent2)
+          , ("B", mBContent, Nothing)
+          ] compile res
+
+  let testWithFailure2 fn name (mAContent1, mAContent2, mBContent) res =
+        testN fn name
+          [ ("A", mAContent1, Just mAContent2)
+          , ("B", mBContent, Nothing)
+          ] compileAllowingFailures res
+
+  let test3 fn name  (mAContent1, mAContent2, mBContent, mCContent) res =
+        testN fn name
+          [ ("A", mAContent1, Just mAContent2)
+          , ("B", mBContent, Nothing)
+          , ("C", mCContent, Nothing)
+          ] compile res
+
+  let testWithFailure3 fn name (mAContent1, mAContent2, mBContent, mCContent) res =
+        testN fn name
+          [ ("A", mAContent1, Just mAContent2)
+          , ("B", mBContent, Nothing)
+          , ("C", mCContent, Nothing)
+          ] compileAllowingFailures res
+
+  let recompile2 fn name ms =
+        test2 fn ("recompiles when upstream changed effectively: " <> name) ms ["A", "B"]
+
+  let recompileWithFailure2 fn name ms =
+        testWithFailure2 fn ("recompiles when upstream changed effectively: " <> name) ms ["A", "B"]
+
+  let noRecompile2 fn name ms =
+        test2 fn ("does not recompile when upstream not changed effectively: " <> name) ms ["A"]
+
   before_ (rimraf modulesDir >> rimraf sourcesDir >> createDirectory sourcesDir) $ do
     it "does not recompile if there are no changes" $ do
-      let modulePath = sourcesDir </> "Module.purs"
+      let mPath = sourcesDir </> "Module.purs"
```

Comment:
This is already a large-ish PR, and I would prefer it if renames like this weren't included, at least as part of this work. (I'm not arguing that one or the other name is better; just that it's not so bad the way that it is that it needs changing.)